homeassistant/components/stt/__init__.py

```diff
@@ -36,33 +49,171 @@
 
 __all__ = [
     "async_get_provider",
+    "async_get_provider_entity",
     "AudioBitRates",
     "AudioChannels",
     "AudioCodecs",
     "AudioFormats",
     "AudioSampleRates",
     "DOMAIN",
     "Provider",
+    "ProviderEntity",
     "SpeechMetadata",
     "SpeechResult",
     "SpeechResultState",
 ]
 
+_LOGGER = logging.getLogger(__name__)
+
+
+@callback
+def async_get_provider_entity(
+    hass: HomeAssistant, entity_id: str
+) -> ProviderEntity | None:
+    """Return provider entity."""
+    component: EntityComponent[ProviderEntity] = hass.data[DOMAIN]
+
+    return component.get_entity(entity_id)
+
 
 async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
     """Set up STT."""
+    component = hass.data[DOMAIN] = EntityComponent[ProviderEntity](
+        _LOGGER, DOMAIN, hass
+    )
+
+    # Cancel the retry setup timer on shutdown (temporary workaround)
+    hass.bus.async_listen_once(
+        EVENT_HOMEASSISTANT_STOP,
+        component._async_shutdown,  # pylint: disable=protected-access
+    )
+
     platform_setups = async_setup_legacy(hass, config)
 
     if platform_setups:
         await asyncio.wait([asyncio.create_task(setup) for setup in platform_setups])
 
-    hass.http.register_view(SpeechToTextView(hass.data[DOMAIN]))
+    hass.http.register_view(SpeechToTextView(hass.data[DATA_PROVIDERS]))
     return True
 
 
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up a config entry."""
+    component: EntityComponent[ProviderEntity] = hass.data[DOMAIN]
+    return await component.async_setup_entry(entry)
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    component: EntityComponent[ProviderEntity] = hass.data[DOMAIN]
+    return await component.async_unload_entry(entry)
+
+
+class ProviderEntity(RestoreEntity):
+    """Represent a single STT provider."""
+
+    _attr_should_poll = False
+    __last_processed: str | None = None
+
+    @property
+    @final
+    def name(self) -> str:
+        """Return the name of the provider entity."""
+        # Only one entity is allowed per platform for now.
+        if self.platform is None:
+            raise RuntimeError("Entity is not added to hass yet.")
```

Comment:
When can this happen?
Related PR: https://github.com/home-assistant/core/pull/88321

Reply:
It can happen if the property is accessed before the entity has been added to Home Assistant.

Reply:
Yes, of course, but when can that happen?

Reply:
A platform could access it after creating the entity.

It's the same for `Entity.hass`. We can decide that we don't want to type `Entity.platform` like it's optional and type it like it's always available although that's not really true. But that's a separate discussion and PR. Here I'm just following the current situation.